widgetssdk/src/main/java/com/glia/widgets/core/queue/QueueMonitor.kt

```diff
@@ -0,0 +1,82 @@
+import com.glia.androidsdk.GliaException
+import com.glia.androidsdk.queuing.Queue
+import com.glia.widgets.core.queue.domain.FetchQueuesUseCase
+import com.glia.widgets.core.queue.domain.SubscribeToQueueUpdatesUseCase
+import com.glia.widgets.core.queue.domain.UnsubscribeFromQueueUpdatesUseCase
+import com.glia.widgets.helper.Data
+import com.glia.widgets.launcher.ConfigurationManager
+import io.reactivex.rxjava3.core.Flowable
+import io.reactivex.rxjava3.processors.BehaviorProcessor
+import java.util.function.Consumer
+
+internal class QueueMonitor(
+    configurationManager: ConfigurationManager,
+    private val fetchQueuesUseCase: FetchQueuesUseCase,
+    private val subscribeToQueueUpdatesUseCase: SubscribeToQueueUpdatesUseCase,
+    private val unsubscribeFromQueueUpdatesUseCase: UnsubscribeFromQueueUpdatesUseCase
+) {
+    // All the queues from the site
+    private val _siteQueues: BehaviorProcessor<List<Queue>> = BehaviorProcessor.create()
+
+    // Queues received from an integrator or default queues
+    private val _integratorQueues: Flowable<List<Queue>> = Flowable.combineLatest(
+        _siteQueues,
+        configurationManager.queueIdsObservable
+    ) { allQueues, queueIds ->
+        mapCurrentQueues(allQueues, queueIds)
+    }
+
+    private val _observableIntegratorQueues: BehaviorProcessor<List<Queue>> = BehaviorProcessor.create()
+    val observableIntegratorQueues = _observableIntegratorQueues.hide()
+
+    private val queueUpdateCallback: Consumer<Queue> = Consumer {
+        updateQueuesList(it)
+    }
+
+    init {
+        fetchSiteQueues()
+    }
+
+    private fun fetchSiteQueues() {
+        val callback: (Array<Queue>?, GliaException?) -> Unit = { queues, error ->
+            if (error == null && !queues.isNullOrEmpty()) {
+                _siteQueues.onNext(queues.toList())
+                subscribeToQueueUpdates()
+            }
+        }
+        fetchQueuesUseCase(callback)
+    }
+
+
+    @SuppressLint("CheckResult")
+    private fun subscribeToQueueUpdates() {
+        _integratorQueues.subscribe { queues ->
+            _observableIntegratorQueues.onNext(queues)
+            unsubscribeFromQueueUpdatesUseCase(null, queueUpdateCallback)
+            subscribeToQueueUpdatesUseCase(queues.map { it.id }.toTypedArray(), {}, queueUpdateCallback)
+        }
+    }
+
+    private fun updateQueuesList(queue: Queue) {
+        val updatedQueues = _observableIntegratorQueues.value?.run {
+            map { if (it.id == queue.id) queue else it }
+        } ?: listOf(queue)
+
+        _observableIntegratorQueues.onNext(updatedQueues)
+    }
+
+    private fun mapCurrentQueues(siteQueues: List<Queue>, queueIds: Data<List<String>>): List<Queue> {
+        return when (queueIds) {
+            is Data.Value -> {
+                siteQueues.filter { queue -> queueIds.result.contains(queue.id) }
+            }
+
+            else -> {
+                siteQueues.filter { it.isDefault == true }
+            }
```

Comment:
I think it would be nice if you use in both `when` cases consistent approach:

```suggestion
            is Data.Value -> {
                siteQueues.filter { queueIds.result.contains(it.id) }
            }

            else -> {
                siteQueues.filter { it.isDefault == true }
            }
```

OR

```suggestion
            is Data.Value -> {
                siteQueues.filter { queue -> queueIds.result.contains(queue.id) }
            }

            else -> {
                siteQueues.filter { queue -> queue.isDefault == true }
            }
```

Reply:
Good catch, thank you.

Reply:
Fixed fe69ca30d1fd76534e309d3d92e65eea8a270bd0